### PR TITLE
refactor(PublicFormRoutes): duplicate /publicform to /api/v3 endpoints

### LINKED ISSUE #1513: [API] Duplicate public form related endpoints to their new /api counterparts
This issue tracks the straight duplication of the various public form endpoints to their `/api` rooted endpoints to follow a RESTful hierarchy. 

The old endpoint should not be deleted for backwards compatibility (at least for a few weeks after the new endpoint has been released).

The current to new mapping is shown below:

\# | Existing endpoints | New endpoints<br>(note that all routes are prefixed with /api/v3) | Description
-- | -- | --|--
1 | GET /:formId/publicform | GET /forms/:formId | retrieve public form data 
2 | POST /:formId/feedback | POST /forms/:formId/feedback | submit form feedback
3 | POST /v2/submissions/encrypt/:formId | POST /forms/:formId/submissions/encrypt | submit storage mode form
4 | POST /v2/submissions/email/:formId | POST /forms/:formId/submissions/email | submit email mode form

### PR DESCRIPTION
**Pull request by seaerchin**
## Problem
duplicates js endpoints in `public-forms.server.routes` to their ts counterparts 
Closes #1513.

## Solution

Duplicated over existing endpoints for `/forms/:formId/feedback` and `/forms/:formId/submissions/email` to the new routes file.

## Tests

1. ported over integration tests for `:/formId/submissions/email`
2. ported over integration tests for `:/formId/submissions/encrypt`
3. ported over integration tests for `:/formId`
4. wrote integration tests for `/:formId/feedback`

## Manual Tests 
**NOTE: Attempt the below tests in all auth types**

- [ ] Create a normal (NIL Auth) email mode form with a few fields and publish it. Open the form and fill in the form together with feedback. The responses should show up in email while the feedback should show up in the admin panel
- [ ] Create a SP email mode form with a few fields and publish it. Open the form and fill in the form together with feedback. The responses should show up in email while the feedback should show up in the admin panel
- [ ] Create a CP email mode form with a few fields and publish it. Open the form and fill in the form together with feedback. The responses should show up in email while the feedback should show up in the admin panel
- [ ] Create a MyInfo email mode form with a few fields and publish it. Open the form and fill in the form together with feedback. The responses should show up in email while the feedback should show up in the admin panel
- [ ] Create a normal (NIL Auth) storage mode form with a few fields and publish it. Open the form and fill in the form together with feedback. The responses should show up in email while the feedback should show up in the admin panel
- [ ] Create a SP storage mode form with a few fields and publish it. Open the form and fill in the form together with feedback. The responses should show up in email while the feedback should show up in the admin panel
- [ ] Create a CP storage mode form with a few fields and publish it. Open the form and fill in the form together with feedback. The responses should show up in email while the feedback should show up in the admin panel
- [ ] Create a MyInfo storage mode form with a few fields and publish it. Open the form and fill in the form together with feedback. The responses should show up in email while the feedback should show up in the admin panel

**MyInfo-specific**

- [ ] 1. Create an email mode form and publish it. When signing in using MockPass, deny authentication. The form should not have been signed in and there should be an error message with MyInfo error displayed

### CODE REVIEW

File: src/app/modules/form/public-form/public-form.controller.ts
Comment (karrui) [inline src/app/modules/form/public-form/public-form.controller.ts:5]: use relative imports instead?

File: src/app/modules/form/public-form/public-form.controller.ts
Comment (karrui) [inline src/app/modules/form/public-form/public-form.controller.ts:30]: nit: prefer if the return type is typed to something like `PrivateFormErrorDto`

File: src/app/routes/api/v3/public/__tests__/public-forms.routes.spec.ts
Comment (karrui) [inline src/app/routes/api/v3/public/__tests__/public-forms.routes.spec.ts:71]: this probably does not matter anymore, can move the import to its correct position and remove the comment and lint ignore line

**Review by karrui — CHANGES_REQUESTED** [review]
you might have forgotten to change all client api calls to the old endpoint to call the new endpoint...

File: src/app/modules/form/public-form/public-form.controller.ts
Comment (seaerchin) [reply src/app/modules/form/public-form/public-form.controller.ts:30]: i agree - waiting on anty's review on [this pr](https://github.com/opengovsg/FormSG/pull/1572/files), which has the dto and will fix once the other pr is merged.

**Comment by seaerchin** [comment]
> you might have forgotten to change all client api calls to the old endpoint to call the new endpoint...

just to clarify, in the **notes** section, i wrote that i didn't change due to what was specified on #1513; do we want to do so immediately or am i mistaken on what #1513 states? (because we should not delete the old endpoints by changing client to call the new ones)

File: src/app/routes/api/v3/public/__tests__/public-forms.routes.spec.ts
Comment (seaerchin) [reply src/app/routes/api/v3/public/__tests__/public-forms.routes.spec.ts:71]: this does matter and removing the import causes the successful `MyInfo` test to fail. This is because the `mockExtractUinFin`, which needs to be setup in order for the test to pass successfully.

**Comment by karrui** [comment]
> (because we should not delete the old endpoints by changing client to call the new ones)

Clarified IRL; call new endpoints with the client (without removing old endpoints); if not we will never be able to delete the old endpoints 😆

File: src/public/modules/forms/services/form-api.client.factory.js
Comment (seaerchin) [inline src/public/modules/forms/services/form-api.client.factory.js:81]: @karrui did this for visibility rather than just changing '/:accessMode' to be `''` since we are going to port over all these to `react-query` eventually.

File: src/app/routes/api/v3/public/public-forms.routes.ts
Comment (mantariksh) [inline src/app/routes/api/v3/public/public-forms.routes.ts:101]: for consistency
```suggestion
  '/:formId([a-fA-F0-9]{24})',
```

File: src/app/routes/api/v3/public/public-forms.routes.ts
Comment (mantariksh) [inline src/app/routes/api/v3/public/public-forms.routes.ts:39]: what do you think about moving this validation middleware into PublicFormController, and exporting both the validation middleware and `handleSubmitFeedback` as an array of middleware instead of separately?

we actually decided to practice this a while ago, but we haven't been very good about enforcing it. it makes sense because `handleSubmitFeedback` assumes that the request has a certain shape, and can only do so safely when the validation middleware comes before it. so the validation and the handler must always go together, so it makes sense to export them as a single unit.

File: src/public/modules/forms/services/form-api.client.factory.js
Comment (mantariksh) [inline src/public/modules/forms/services/form-api.client.factory.js:105]: you can delete the line immediately after this one, since we don't need the `accessMode` param anymore for this

File: src/app/routes/api/v3/public/public-forms.routes.ts
Comment (seaerchin) [reply src/app/routes/api/v3/public/public-forms.routes.ts:39]: do you mean something along the lines of 

```ts
const handleSubmitFeedback = ... (same as current) 
export const handleSubmitFeedbackWithValidation = [celebrate(...), handleSubmitFeedback]
```

Comment (seaerchin) [reply src/app/routes/api/v3/public/public-forms.routes.ts:39]: @mantariksh following up on this - i will collapse 2 other separate endpoints as well 

1. handleEncryptedSubmission
2. handleEmailSubmission 

the only visible middleware should be the rate limiter. This is because we want to always validate captcha and submission parameters before deferring to the respective controllers but limit rate is tweakable and should not be viewed as being crucial to handling the respective submissions.

feel free to correct me if i'm wrong on this!

Comment (mantariksh) [reply src/app/routes/api/v3/public/public-forms.routes.ts]: what about `receiveEmailSubmission`?

Comment (seaerchin) [reply src/app/routes/api/v3/public/public-forms.routes.ts]: hidden from routes and is exported as part of `handleEmailSubmission`

Comment (mantariksh) [reply src/app/routes/api/v3/public/public-forms.routes.ts]: excellent!

**Review by mantariksh — COMMENTED** [review]
lgtm, but could you test on staging before approval?

**Comment by seaerchin** [comment]
@mantariksh will add manual tests in also but it'll be happy path tests because this is a duplication

**Review by mantariksh — APPROVED** [review]

**Review by karrui — APPROVED** [review]

**Pull request merged**
